proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md

```diff
@@ -0,0 +1,221 @@
+However, some developers might consider this an unacceptable risk and prefer another way to get a VDB that cannot accidentally be used to get packages.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+#### auditSources in nuget.config
+
+*NuGet.Config* files have a new section `<auditSources>` where the URL(s) of any audit sources can be specified, which will be used only to download a VDB, and will not be used for downloading packages.
+Like `<packageSources>`, `<auditSources>` can contain zero, one, or more sources, and supports `<clear />` to remove any `auditSources` inherited from parent *nuget.config* files.
+When there is at least one source in `auditSources`, NuGet should no longer use `packageSources` to get a VDB.
+Omitting `<clear />` will not use `packageSources` in addition to `auditSources`.
+
+For example, consider a developer at Contoso using a company-internal source as a single package source, and using nuget.org as an audit source:
+
+```xml
+<configuration>
+  <packageSources>
+    <!-- Clear to ensure package sources not inherited from other config files -->
+    <clear />
+    <add key="contoso" value="https://internal.corp.contoso.com/packages/nuget/index.json" >
+  </packageSources>
+  <auditSources>
+    <!-- Clear to ensure audit sources not inherited from other config files -->
+    <clear />
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" >
+  </auditSources>
+</configuration>
+```
+
+Any URL specified in the `auditSources` section must implement the [NuGet Server API V3 protocol](https://learn.microsoft.com/nuget/api/overview).
+However, the server will only need to implement the [`VulnerabilityInfo` resource](https://learn.microsoft.com/en-us/nuget/api/vulnerability-info) (and service index), and can avoid implementing all the other resources that are required to be used as a package source.
+
+NuGet will use same semantics for `auditSources` as it already has with `packageSources`.
+This means that `auditSources` are accumulated from other nuget.config files, and `<clear />` can be used to remove them.
+`auditSources` will also follow `packageSources` with regards to insecure http warnings, errors, and `allowInsecureConnections` configurations.
+Sources that need credentials will use `packageSourceCredentials`, using the `key` as the lookup, just as `packageSources` does.
+This means that credentials can be re-used by using the same key for a source that is defined in both `packageSources` and `auditSources`.
```

Comment:
I'm also curious about the combinations of package sources and audit sources but maybe it's a little different from what Anand and Jon are asking about in [their comment thread](https://github.com/NuGet/Home/pull/12918/files#r1372141490). The combination that I think is most interesting to focus on is when there a no audit sources defined and 1 or more package sources with no VDB. When a user with this config updates to the client version that carries this change, how will they be guided to adding NuGet.org as the audit source (which seems like a very good suggestion for someone to give them)? Or is there a default where we inject NuGet.org as `<auditSources>` perhaps at the user level allowing them to override that with a audit sources `<clear />` at the repo level?

This is maybe along the same lines of [Svetlana's comment](https://github.com/NuGet/Home/pull/12918/files#r1376881089) but I thought she was more asking about how new NuGet.config files get initialized and less about adoption of this feature by existing nuget.configs.

Reply:
IMO the logic should be.

1. If you have a package source with a VDB, it is implicitly added to the auditsources (not actually in the nuget.config though)
2. If you don't have a package source with a VDB, you explicitly add it to the auditsources in the nuget.config.
3. If you have multiple auditsources defined, maybe there is a question of whether they audit their own sources accordingly or if one should gesture what source to audit for all packages? I think we are seeing this from the central registry perspective today but in the far future this could be used in different contexts. 

Apologies as I'm catching up.

Reply:
Jon, I interpret your comment as being orthogonal to Joel's question.

> The combination that I think is most interesting to focus on is when there a no audit sources defined and 1 or more package sources with no VDB.

I think having at least one source that doesn't provide a VDB is going to be very common, and would be too spammy to have any output for. Also probably wouldn't be productive if a different source does provide a VDB.

However, a restore where none of the sources provided a VDB is something I think can be reasonable to provide feedback on. Currently, I'm thinking maybe the end of restore summary could have a message something like:

```text
Audit sources used:
   None, see https://aka.ms/nuget/audit#sources for more information
```

The part of the spec that proposes the restore summary change says that if audit isn't run, it shouldn't output this part of the message. However, maybe we should only omit this message if Audit is disabled, rather than no audit sources used.

However, an issue with this is that this message is only output at normal verbosity, and both VS and the `dotnet` CLI default to minimal verbosity, so many customers won't see the message.

> Or is there a default where we inject NuGet.org as <auditSources> perhaps at the user level allowing them to override that with a audit sources <clear /> at the repo level?

That's something I proposed as an idea earlier, but other people in the team didn't like it. I still think it's a good idea, and in line with what all other package manager ecosystems are doing. 🤷 Frankly, I think it's a better for package sources as well, compared to "create default config if it doesn't already exist", and would have avoided all the issues with had with the config tracking file and auto-adding nuget.org in the past.